compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -216,6 +215,24 @@ impl<'f> LoopInvariantContext<'f> {
         is_loop_invariant && can_be_deduplicated
     }
 
+    /// Keep track of a loop induction variable and respective upper bound.
+    /// In the case of a nested loop, this will be used by later loops to determine
+    /// whether they have operations reliant upon the maximum induction variable.
+    /// When within the current loop, the known upper bound can be used to simplify instructions,
+    /// such as transforming a checked add to an unchecked add.
+    fn set_induction_var_bounds(&mut self, loop_: &Loop, current_loop: bool) {
+        let upper_bound = loop_.get_const_upper_bound(self.inserter.function);
+        if let Some(upper_bound) = upper_bound {
+            let induction_variable = loop_.get_induction_variable(self.inserter.function);
+            let induction_variable = self.inserter.resolve(induction_variable);
+            if current_loop {
+                self.current_induction_variables.insert(induction_variable, upper_bound);
+            } else {
+                self.outer_induction_variables.insert(induction_variable, upper_bound);
+            }
```

Comment:
Seems like these two maps will only differ by the current loop's induction variable (we never remove anything from `current_induction_variables`).

We could then simplify this to just have `induction_variables` and then track the current induction variable if we need to special case something.

Reply:
Yeah I had thought about combining them as well, but was unsure how much it would simplify things. Tracking the current induction var would require also special casing this when checking whether we can evaluate a binary operation for hoisting rather than simply passing `self.outer_induction_variables`. I ultimately just went with an extra map as there shouldn't be very many induction variables. If we feel one `induction_variables` map would be simpler though I can switch to this.

Reply:
I think at the very least we want to have some documentation near the definitions of these two maps making this relationship between them very explicit and why we've chosen to do this. Merging them can be left as a TODO.

Reply:
I switched to clearing the `current_induction_variables` map for every loop. And have added some comments.